### PR TITLE
Show fwlink on HTTPS certificate errors (#83)

### PR DESCRIPTION
**Pull request by cesarbs**
#83

### CODE REVIEW

File: src/Microsoft.AspNetCore/CertificateConfigurationException.cs [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore
+{
+    internal class CertificateConfigurationException : Exception
+    {
+        public CertificateConfigurationException(string message)
+            : base(message + ". For information on configuring HTTPS see https://go.microsoft.com/fwlink/?linkid=848054.")
```
Comment (muratg) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: Why not `$"{message}. For infor...."`?

Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: Why is it adding a period to the previous message at all?

Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: O I C. All the previous exception messages should be complete sentences (wherever possible) and end with periods... Though some of them I can see why they don't...

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 2]
```diff
@@ -99,7 +99,7 @@ private X509Certificate2 LoadSingle(string certificateName)
 
             if (!certificateConfiguration.Exists())
             {
-                throw new InvalidOperationException($"No certificate named {certificateName} found in configuration");
+                throw new CertificateConfigurationException($"No certificate named {certificateName} found in configuration for the current environment");
```
Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: This sentence needs to end with a period.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 3]
```diff
@@ -119,7 +119,7 @@ private X509Certificate2 LoadSingle(IConfigurationSection certificateConfigurati
                     certificateSource = new CertificateStoreSource(_certificateStoreLoader);
                     break;
                 default:
-                    throw new InvalidOperationException($"Invalid certificate source kind: {sourceKind}");
+                    throw new CertificateConfigurationException($"Invalid certificate source kind: {sourceKind}");
```
Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: Change this to be a sentence: `Invalid certificate source kind '{sourceKind}'.`

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 4]
```diff
@@ -163,7 +163,7 @@ public override X509Certificate2 Load()
 
                 if (error != null)
                 {
-                    throw error;
+                    throw new CertificateConfigurationException($"Unable to load certificate from file '{Path}': {error.Message}");
```
Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: `Unable to load certificate from file '{Path}'. Error details: '{error.Message}'.`

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 5]
```diff
@@ -203,7 +203,7 @@ public override X509Certificate2 Load()
             {
                 if (!Enum.TryParse(StoreLocation, ignoreCase: true, result: out StoreLocation storeLocation))
                 {
-                    throw new InvalidOperationException($"Invalid store location: {StoreLocation}");
+                    throw new CertificateConfigurationException($"Invalid certificate store location: {StoreLocation}");
```
Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: `The certificate store location '{StoreLocation}' is invalid.`

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 6]
```diff
@@ -63,7 +63,7 @@ private void BindConfiguration(KestrelServerOptions options)
 
                     if (certificate == null)
                     {
-                        throw new InvalidOperationException($"Unable to load certificate for endpoint '{endPoint.Key}'");
+                        throw new CertificateConfigurationException($"Unable to load certificate for endpoint '{endPoint.Key}'");
```
Comment (Eilon) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: End with period.

File: src/Microsoft.AspNetCore/CertificateConfigurationException.cs (hunk 1 above)
Comment (Eilon) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: Actually I suggested fixes for all the exception messages below. Then remove the leading period on this one here.

**Comment by danroth27** [comment]
The CertificateLoader will be used for other scenarios than loading the HTTPS certificate. We're using it to load signing certificates as well. So I don't think the information about how to enable HTTPS is being added in the right context. I.e. you don't need instructions about setting up HTTPS if the configuration for your signing certificate is invalid. I think the HTTPS information needs to be added when setting up HTTPS with Kestrel.

Also, we should log a warning if the CertificateLoader try to find a certificate with a particular subject name in a store but doesn't find any.

File: src/Microsoft.AspNetCore/CertificateLoader.cs [hunk 7]
```diff
@@ -99,7 +125,7 @@ private X509Certificate2 LoadSingle(string certificateName)
 
             if (!certificateConfiguration.Exists())
             {
-                throw new InvalidOperationException($"No certificate named {certificateName} found in configuration");
+                throw new KeyNotFoundException($"No certificate named '{certificateName}' found in configuration for the current environment.");
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: @Tratcher Is it ok to throw this exception type here?

File: src/Microsoft.AspNetCore/CertificateConfigurationException.cs [hunk 8]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+
+namespace Microsoft.AspNetCore
+{
+    internal class CertificateConfigurationException : Exception
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: @Tratcher Is it ok to keep this internal?

File: test/Microsoft.AspNetCore.FunctionalTests/CertificateLoaderTests.cs [hunk 9]
```diff
@@ -988,12 +1013,78 @@ public void ReturnsNull_MultipleCertificatesInline_FileAndStore_NotFoundInStore(
...
+                null,
+                loggerFactory.Object,
+                Mock.Of<ICertificateFileLoader>(),
+                Mock.Of<ICertificateStoreLoader>());
+
+            var loadedCertificates = certificateLoader.Load(configuration.GetSection("TestConfig:Certificates"));
+            Assert.Equal(0, loadedCertificates.Count());
+            Assert.Single(logger.LogMessages, logMessage =>
+                logMessage.LogLevel == LogLevel.Warning &&
+                logMessage.Message == "Unable to find a matching certificate for subject 'localhost' in store 'My' in 'CurrentUser'.");
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.FunctionalTests/CertificateLoaderTests.cs]: @Eilon Is the language alright? Should log messages also end with a period?

File: src/Microsoft.AspNetCore/CertificateLoader.cs (hunk 7 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateLoader.cs]: 👍

File: src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs [hunk 10]
```diff
@@ -27,7 +31,7 @@ public void Configure(KestrelServerOptions options)
 
         private void BindConfiguration(KestrelServerOptions options)
         {
-            var certificateLoader = new CertificateLoader(_configurationRoot.GetSection("Certificates"));
+            var certificateLoader = new CertificateLoader(_configurationRoot.GetSection("Certificates"), _loggerFactory);
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore/KestrelServerOptionsSetup.cs]: @danroth27 I think we should put `CertificateLoader` in DI.

@davidfowl What do you think?

File: src/Microsoft.AspNetCore/CertificateConfigurationException.cs (hunk 8 above)
Comment (Tratcher) [inline src/Microsoft.AspNetCore/CertificateConfigurationException.cs]: :-1:

**Pull request merged**